packages/jsandbox/buildscript/src/main.rs

```diff
@@ -0,0 +1,294 @@
+        Ok(self.workspace()?.join("target"))
+    }
+
+    fn read_metadata(&mut self) -> Result<()> {
+        let manifest = cargo_toml::Manifest::from_path(self.cargo_toml_path()?)?;
+        self.cargo_manifest = Some(manifest);
+        Ok(())
+    }
+
+    async fn project_name(&self) -> Result<&str> {
+        let cargo_path = self.cargo_toml_path()?;
+        Ok(self
+            .cargo_manifest
+            .as_ref()
+            .ok_or_else(|| anyhow!("{} metadata hasn’t been read", cargo_path.display()))?
+            .package
+            .as_ref()
+            .ok_or_else(|| anyhow!("{} must have a [package] section", cargo_path.display()))?
+            .name
+            .as_str())
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let mut args = CliArgs::parse();
+    args.read_metadata()?;
+    build_jsandbox(&args).await.context("jsandbox")?;
+
+    Ok(())
+}
+
+macro_rules! promise_sequence {
+    ($a:expr, $b:expr) => {
+        async move {
+            match $a.await {
+                Ok(_) => $b.await,
+                v @ Err(_) => v,
+            }
+        }
+    };
+}
+
+async fn build_jsandbox(args: &CliArgs) -> Result<()> {
+    build_crate(args, "jsandbox")
+        .await
+        .context("Building crate")?;
+
+    run_bindgen(args, "jsandbox.wasm")
+        .await
+        .context("Running bindgen")?;
+    tokio::try_join!(
+        async {
+            fix_bindgen_js(args, "jsandbox_bg.js")
+                .await
+                .context("Fixing bindgen JS")
+        },
+        async {
+            run_wasm_opt(
+                args,
+                args.target_path()?
+                    .join("wasm-bindgen")
+                    .join("jsandbox_bg.wasm"),
+            )
+            .await
+            .context("Processing bindgen wasm")
+        }
+    )?;
+    Ok(())
+}
+
+async fn build_crate(args: &CliArgs, name: impl AsRef<str>) -> Result<(), anyhow::Error> {
+    let mut build_args: Vec<_> = vec!["build", "--target", "wasm32-wasi", "-p", name.as_ref()];
+    if args.release {
+        build_args.push("--release");
+    }
+    let mut build_command = Command::new("cargo").args(build_args).spawn()?;
+    build_command.wait().await?;
+    Ok(())
+}
+
+async fn run_bindgen(args: &CliArgs, name: impl AsRef<str>) -> Result<()> {
+    let build_mode = if args.release { "release" } else { "debug" };
+    let input_path = PathBuf::from("./target/wasm32-wasi/")
+        .join(build_mode)
+        .join(name.as_ref());
+    let mut bindgen = wasm_bindgen_cli_support::Bindgen::new();
+    bindgen
+        .input_path(input_path)
+        .keep_debug(args.release)
+        .bundler(true)?
+        .emit_start(true)
+        .reference_types(true)
+        .omit_default_module_path(false);
+    let output = args.target_path()?.join("wasm-bindgen");
+    tokio::fs::create_dir_all(&output).await?;
+    bindgen.generate(&output)?;
+    Ok(())
+}
+
+// FIXME: Brittle AF
```

Comment:
Ah good shout. I copied this over from another project and didn't really reread what I am doing here 😅 

By default, the JS output of wasm-bindgen creates a JS modules that contains a single WAsm module instance. As a result, you will only ever have a single instance of your WebAssembly instance. That is in line with Wasm's ES module integration proposal, but really not a great mode for something like this, where you'd probably want to create one instance per sandbox that you create. 

The point of this "fixing" is to rewrite the generated JS to allow the creation of multiple instances of the Wasm module. To that end, I remove the `export` from all functions, and wrap the entire module in a new exported IIFE that acts as a factory function. The detection of which functions are exported is a simple search/replace, hence why I marked it as brittle. It works, and I think it's fairly reliable considering what code wasm-bindgen generates, but at some point you'd probably use a proper ECMAScript parser to do this rewriting rather than string-based search/replace.